File: gear_sonic_deploy/src/g1/g1_deploy_onnx_ref/tests/sonic_trt_inference_bridge.cpp
#include <algorithm>
#include <cerrno>
#include <cstdint>
#include <cstring>
#include <iostream>
#include <string>
#include <vector>

#include <unistd.h>

#include "../include/control_policy.hpp"
#include "../include/encoder.hpp"

namespace {

constexpr std::uint32_t kReadyMagic = 0x31545254;  // "TRT1" little-endian
constexpr std::uint32_t kProtocolVersion = 1;

bool ReadExact(int fd, void* data, std::size_t bytes) {
  auto* ptr = static_cast<unsigned char*>(data);
  while (bytes > 0) {
    const ssize_t n = ::read(fd, ptr, bytes);
    if (n == 0) {
      return false;
    }
    if (n < 0) {
      if (errno == EINTR) {
        continue;
      }
      return false;
    }
    ptr += n;
    bytes -= static_cast<std::size_t>(n);
  }
  return true;
}

bool WriteExact(int fd, const void* data, std::size_t bytes) {
  const auto* ptr = static_cast<const unsigned char*>(data);
  while (bytes > 0) {
    const ssize_t n = ::write(fd, ptr, bytes);
    if (n < 0) {
      if (errno == EINTR) {
        continue;
      }
      return false;
    }
    ptr += n;
    bytes -= static_cast<std::size_t>(n);
  }
  return true;
}

bool WriteU32(int fd, std::uint32_t value) {
  return WriteExact(fd, &value, sizeof(value));
}

bool ReadU32(int fd, std::uint32_t& value) {
  return ReadExact(fd, &value, sizeof(value));
}

void Usage(const char* argv0) {
  std::cerr
      << "Usage: " << argv0
      << " --encoder <model_encoder.onnx> --decoder <model_decoder.onnx>"
      << " --response_fd <fd> [--encoder_fp16] [--decoder_fp16]\n";
}

}  // namespace

int main(int argc, char** argv) {
  std::string encoder_path;
  std::string decoder_path;
  int response_fd = -1;
  bool encoder_fp16 = false;
  bool decoder_fp16 = false;

  for (int i = 1; i < argc; ++i) {
    const std::string arg = argv[i];
    if (arg == "--encoder" && i + 1 < argc) {
      encoder_path = argv[++i];
    } else if (arg == "--decoder" && i + 1 < argc) {
      decoder_path = argv[++i];
    } else if (arg == "--response_fd" && i + 1 < argc) {
      response_fd = std::stoi(argv[++i]);
    } else if (arg == "--encoder_fp16") {
      encoder_fp16 = true;
    } else if (arg == "--decoder_fp16") {
      decoder_fp16 = true;
    } else if (arg == "-h" || arg == "--help") {
      Usage(argv[0]);
      return 0;
    } else {
      std::cerr << "Unknown or incomplete argument: " << arg << "\n";
      Usage(argv[0]);
      return 2;
    }
  }

  if (encoder_path.empty() || decoder_path.empty() || response_fd < 0) {
    Usage(argv[0]);
    return 2;
  }

  EncoderEngine encoder;
  PolicyEngine decoder;
  if (!encoder.Initialize(encoder_path, encoder_fp16)) {
    std::cerr << "Failed to initialize TensorRT encoder.\n";
    return 1;
  }
  if (!decoder.Initialize(decoder_path, decoder_fp16)) {
    std::cerr << "Failed to initialize TensorRT decoder.\n";
    return 1;
  }

  const std::uint32_t enc_in_dim = static_cast<std::uint32_t>(encoder.GetInputDimension());
  const std::uint32_t token_dim = static_cast<std::uint32_t>(encoder.GetTokenDimension());
  const std::uint32_t dec_in_dim = static_cast<std::uint32_t>(decoder.GetInputDimension());
  const std::uint32_t action_dim = static_cast<std::uint32_t>(decoder.GetActionDimension());

  if (!WriteU32(response_fd, kReadyMagic) ||
      !WriteU32(response_fd, kProtocolVersion) ||
      !WriteU32(response_fd, enc_in_dim) ||
      !WriteU32(response_fd, token_dim) ||
      !WriteU32(response_fd, dec_in_dim) ||
      !WriteU32(response_fd, action_dim)) {
    std::cerr << "Failed to write bridge ready header.\n";
    return 1;
  }

  while (true) {
    unsigned char request_type = 0;
    if (!ReadExact(STDIN_FILENO, &request_type, sizeof(request_type))) {
      break;
    }
    if (request_type == 'Q') {
      break;
    }

    std::uint32_t input_count = 0;
    if (!ReadU32(STDIN_FILENO, input_count)) {
      std::cerr << "Failed to read request input count.\n";
      return 1;
    }

    std::vector<float> input(input_count);
    if (!ReadExact(STDIN_FILENO, input.data(), input.size() * sizeof(float))) {
      std::cerr << "Failed to read request input tensor.\n";
      return 1;
    }

    if (request_type == 'E') {
      if (input_count != enc_in_dim) {
        std::cerr << "Encoder input count mismatch: got " << input_count
                  << ", expected " << enc_in_dim << "\n";
        return 1;
      }
      auto& input_buffer = encoder.GetInputBuffer();
      std::copy(input.begin(), input.end(), input_buffer.begin());
      if (!encoder.Encode()) {
        std::cerr << "Encoder inference failed.\n";
        return 1;
      }
      const auto& output = encoder.GetTokenBuffer();
      const unsigned char response_type = 'E';
      if (!WriteExact(response_fd, &response_type, sizeof(response_type)) ||
          !WriteU32(response_fd, token_dim) ||
          !WriteExact(response_fd, output.data(), output.size() * sizeof(float))) {
        std::cerr << "Failed to write encoder response.\n";
        return 1;
      }
    } else if (request_type == 'D') {
      if (input_count != dec_in_dim) {
        std::cerr << "Decoder input count mismatch: got " << input_count
                  << ", expected " << dec_in_dim << "\n";
        return 1;
      }
      auto& input_buffer = decoder.GetInputBuffer();
      std::copy(input.begin(), input.end(), input_buffer.begin());
      if (!decoder.Infer()) {
        std::cerr << "Decoder inference failed.\n";
        return 1;
      }
      const auto& output = decoder.GetActionBuffer();
      const unsigned char response_type = 'D';
      if (!WriteExact(response_fd, &response_type, sizeof(response_type)) ||
          !WriteU32(response_fd, action_dim) ||
          !WriteExact(response_fd, output.data(), output.size() * sizeof(float))) {
        std::cerr << "Failed to write decoder response.\n";
        return 1;
      }
    } else {
      std::cerr << "Unknown request type: " << static_cast<int>(request_type) << "\n";
      return 1;
    }
  }

  return 0;
}

File: gear_sonic/scripts/sonic_trt_inference_bridge.py
"""Python client for the standalone C++ TensorRT SONIC inference bridge."""

from __future__ import annotations

import os
import struct
import subprocess
from pathlib import Path
from typing import Any

import numpy as np


_READY = struct.Struct("<6I")
_REQUEST = struct.Struct("<cI")
_RESPONSE = struct.Struct("<cI")
_READY_MAGIC = 0x31545254
_PROTOCOL_VERSION = 1


def _read_exact(fd: int, size: int) -> bytes:
    chunks: list[bytes] = []
    remaining = size
    while remaining:
        chunk = os.read(fd, remaining)
        if not chunk:
            raise RuntimeError("TensorRT bridge closed before sending a complete response")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _write_exact(stream: Any, data: bytes) -> None:
    stream.write(data)
    stream.flush()


class _TrtSessionAdapter:
    def __init__(self, bridge: "SonicTrtInferenceBridge", request_type: bytes, input_dim: int, output_dim: int):
        self._bridge = bridge
        self._request_type = request_type
        self.input_dim = input_dim
        self.output_dim = output_dim

    def run(self, _outputs: Any, inputs: dict[str, np.ndarray]) -> list[np.ndarray]:
        if "obs_dict" not in inputs:
            raise KeyError("TensorRT bridge expects input key 'obs_dict'")
        obs = np.asarray(inputs["obs_dict"], dtype=np.float32)
        had_batch = obs.ndim == 2
        if had_batch:
            if obs.shape[0] != 1:
                raise ValueError(f"TensorRT bridge only supports batch size 1, got {obs.shape}")
            obs_1d = np.ascontiguousarray(obs[0], dtype=np.float32)
        else:
            obs_1d = np.ascontiguousarray(obs.reshape(-1), dtype=np.float32)

        if obs_1d.size != self.input_dim:
            raise ValueError(f"TensorRT bridge expected {self.input_dim} inputs, got {obs_1d.size}")

        out = self._bridge.infer(self._request_type, obs_1d, self.output_dim)
        return [out[np.newaxis, :] if had_batch else out]


class SonicTrtInferenceBridge:
    """Persistent subprocess bridge to C++ EncoderEngine + PolicyEngine TensorRT."""

    def __init__(
        self,
        encoder_path: str | Path,
        decoder_path: str | Path,
        *,
        executable: str | Path = "gear_sonic_deploy/target/release/sonic_trt_inference_bridge",
        encoder_fp16: bool = False,
        decoder_fp16: bool = False,
    ):
        read_fd, write_fd = os.pipe()
        cmd = [
            str(executable),
            "--encoder",
            str(encoder_path),
            "--decoder",
            str(decoder_path),
            "--response_fd",
            str(write_fd),
        ]
        if encoder_fp16:
            cmd.append("--encoder_fp16")
        if decoder_fp16:
            cmd.append("--decoder_fp16")

        self._read_fd = read_fd
        self._write_fd = write_fd
        self._proc = subprocess.Popen(
            cmd,
            stdin=subprocess.PIPE,
            pass_fds=(write_fd,),
        )
        os.close(write_fd)
        self._write_fd = -1

        ready = _READY.unpack(_read_exact(read_fd, _READY.size))
        magic, version, enc_in_dim, token_dim, dec_in_dim, action_dim = ready
        if magic != _READY_MAGIC or version != _PROTOCOL_VERSION:
            self.close()
            raise RuntimeError(f"Invalid TensorRT bridge ready header: magic={magic:#x} version={version}")

        self.encoder_input_dim = int(enc_in_dim)
        self.token_dim = int(token_dim)
        self.decoder_input_dim = int(dec_in_dim)
        self.action_dim = int(action_dim)
        self.encoder_session = _TrtSessionAdapter(self, b"E", self.encoder_input_dim, self.token_dim)
        self.decoder_session = _TrtSessionAdapter(self, b"D", self.decoder_input_dim, self.action_dim)

    def infer(self, request_type: bytes, obs: np.ndarray, output_dim: int) -> np.ndarray:
        if self._proc.stdin is None:
            raise RuntimeError("TensorRT bridge stdin is closed")
        payload = _REQUEST.pack(request_type, int(obs.size)) + obs.tobytes(order="C")
        _write_exact(self._proc.stdin, payload)
        response_type, count = _RESPONSE.unpack(_read_exact(self._read_fd, _RESPONSE.size))
        if response_type != request_type:
            raise RuntimeError(f"TensorRT bridge response type mismatch: {response_type!r} != {request_type!r}")
        if count != output_dim:
            raise RuntimeError(f"TensorRT bridge output dim mismatch: {count} != {output_dim}")
        raw = _read_exact(self._read_fd, int(count) * np.dtype(np.float32).itemsize)
        return np.frombuffer(raw, dtype=np.float32).copy()

    def close(self) -> None:
        proc = getattr(self, "_proc", None)
        if proc is not None and proc.stdin is not None and proc.poll() is None:
            try:
                _write_exact(proc.stdin, b"Q")
            except Exception:
                pass
            try:
                proc.stdin.close()
            except Exception:
                pass
            try:
                proc.wait(timeout=2.0)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=2.0)
        read_fd = getattr(self, "_read_fd", -1)
        if read_fd >= 0:
            os.close(read_fd)
            self._read_fd = -1

    def __enter__(self) -> "SonicTrtInferenceBridge":
        return self

    def __exit__(self, _exc_type: Any, _exc: Any, _tb: Any) -> None:
        self.close()

    def __del__(self) -> None:
        try:
            self.close()
        except Exception:
            pass
